lib/internal/vm/Module.js

```diff
@@ -0,0 +1,150 @@
+      context = options.context.sandbox;
+      if (options.context[kContextModuleCount] === undefined)
+        throw new errors.TypeError('ERR_CONTEXT_NOT_MODULE_CONTEXT');
+    }
+
+    const url = options.url ||
+      `vm:module(${context !== undefined ?
+        options.context[kContextModuleCount]++ :
+        globalModuleId++})`;
+
+    const wrap = new ModuleWrap(src, url, {
+      context,
+      lineOffset: options.lineOffset,
+      columnOffset: options.columnOffset,
+    });
+
+    Object.defineProperties(this, {
+      [kWrap]: { value: wrap },
+      context: { value: options.context },
+    });
+  }
+
+  get status() {
+    return StatusMap[this[kWrap].getStatus()];
+  }
+
+  get url() {
+    return this[kWrap].url;
+  }
+
+  get namespace() {
+    if (this[kWrap].getStatus() !== kEvaluated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be evaluated');
+    return this[kWrap].namespace();
+  }
+
+  get dependencySpecifiers() {
+    if (this[kDependencyCache] !== undefined)
+      return this[kDependencyCache];
+
+    return this[kDependencyCache] = this[kWrap].getDependencySpecifiers();
+  }
+
+  async link(fn) {
+    const wrap = this[kWrap];
+    if (wrap.isLinked())
+      throw new errors.Error('ERR_VM_MODULE_ALREADY_LINKED');
+    if (wrap.getStatus() !== kUninstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be uninstantiated');
+    const promises = [];
+    wrap.link(async (specifier) => {
+      const p = fn(specifier);
+      promises.push(p);
+      const m = await p;
+      if (!(m instanceof Module))
+        throw new errors.Error('ERR_VM_MODULE_NOT_MODULE');
+      return m[kWrap];
+    });
+    await Promise.all(promises);
+  }
+
+  instantiate() {
+    const wrap = this[kWrap];
+    if (wrap.getStatus() !== kUninstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be uninstantiated');
+    if (!wrap.isLinked())
+      throw new errors.Error('ERR_VM_MODULE_UNLINKED');
+    this[kWrap].instantiate();
+  }
+
+  async evaluate(options) {
+    const status = this[kWrap].getStatus();
+    if (status !== kInstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be instantiated');
+    if (status === kEvaluating ||
+      status === kEvaluated ||
+      status === kErrored)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'is already evaluated');
+    const result = this[kWrap].evaluate(options);
+    return { result, __proto__: null };
+  }
+
+  [util.inspect.custom](depth, options) {
+    return `Module ${util.inspect({
+      status: this.status,
+      url: this.url
+    }, options)}`;
+  }
+}
+
+class ModuleContext {
+  constructor(sandbox) {
+    const context = require('vm').createContext(sandbox);
+    Object.defineProperties(this, {
+      sandbox: {
+        value: context,
+        writable: false,
+        enumerable: true,
+        configurable: false,
+      },
+      [kContextModuleCount]: {
```

Comment:
What is this exposed for? / why not a weakmap?

Reply:
ya tgu and I discussed this last night on irc, got rid of the whole class